R/serializer.R

```diff
@@ -17,30 +8,267 @@ NULL
 #' @param serializer The serializer to be added.
+#' @param verbose Logical value which determines if a message should be printed when overwriting serializers
 #'
 #' @export
-addSerializer <- function(name, serializer){
-  if (is.null(.globals$serializers)){
-    .globals$serializers <- list()
-  }
-  if (!is.null(.globals$serializers[[name]])){
-    stop ("Already have a serializer by the name of ", name)
+addSerializer <- function(name, serializer, verbose = TRUE) {
+  if (!is.null(.globals$serializers[[name]])) {
+    if (isTRUE(verbose)) {
+      warning("Overwriting serializer: ", name)
+    }
   }
   .globals$serializers[[name]] <- serializer
 }
 
+# internal function to use directly within this file only. (performance purposes)
+# Other files should use `serializer_identity()` to avoid confusion
+serializer_identity_ <- function(val, req, res, errorHandler) {
+  tryCatch({
+    res$body <- val
+    res$toResponse()
+  }, error = function(err) {
+    errorHandler(req, res, err)
+  })
+}
+serializer_identity <- function(){
+  serializer_identity_
+}
 
 
-nullSerializer <- function(){
-  function(val, req, res, errorHandler){
+#' Return an attachment response
+#'
+#' This will set the appropriate fields in the `Content-Disposition` header value.
+#' To make sure the attachement is used, be sure your serializer eventually calls `serializer_headers`
+#'
+#' @param value Response value to be saved
+#' @param filename File name to use when saving the attachment.
+#'   If no `filename` is provided, the `value` will be treated as a regular attachment
+#' @return Object with class `"plumber_attachment"`
+#' @examples
+#' # plumber.R
+#'
+#' #' @get /data
+#' #' @serializer csv
+#' function() {
+#'   # will cause the file to be saved as `iris.csv`, not `data` or `data.csv`
+#'   as_attachment(iris, "iris.csv")
+#' }
+as_attachment <- function(value, filename = NULL) {
+  stopifnot(is.character(filename) || is.null(filename))
+  if (is.character(filename)) {
+    stopifnot(length(filename) == 1)
+  }
+  structure(
+    list(
+      value = value,
+      filename = filename
+    ),
+    class = "plumber_attachment"
+  )
+}
+
+
+#' Plumber Serializers
+#'
+#' Serializers are used in Plumber to transform the R object produced by a
+#' filter/endpoint into an HTTP response that can be returned to the client. See
+#' [here](https://book.rplumber.io/articles/rendering-output.html#serializers-1) for
+#' more details on Plumber serializers and how to customize their behavior.
+#' @describeIn serializers Add a static list of headers to each return value. Will add `Content-Disposition` header if a value is the result of `as_attachment()`.
+#' @param ... extra arguments supplied to respective internal serialization function.
+#' @param headers `list()` of headers to add to the response object
+#' @param serialize_fn Function to serialize the data. The result object will be converted to a character string. Ex: `jsonlite::toJSON`.
```

Comment:
Maybe try `[jsonlite::toJSON()]` with recent version of roxygen2

Reply:
`as_attachment` isn't exported 😁